### PR TITLE
Fix precondition in DcpHostService

### PR DESCRIPTION
**Pull request by drewnoakes**
Two lines are changed here, though it's the second (in `StopAsync`) that has a functional change.

The first guard clause bails if:
- the publisher is not null, AND
- the publisher is not "dcp"

The second guard clause bails if:
- the publisher is not null, OR
- the publisher is not "dcp"

The key difference is the AND vs the OR. This looks like a bug to me as the start/stop methods pair up. The stop clause is more permissive, meaning we could try to stop something that was never started.

My interpretation of this code is that both clauses are intended to be the same, and this commit brings them in line with one another.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1943)

### CODE REVIEW

**Comment by JamesNK** [comment]
Can we unit test this to stop it regressing again?

File: src/Aspire.Hosting/Dcp/DcpHostService.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ internal sealed partial class DcpHostService : IHostedLifecycleService, IAsyncDi
 
     public async Task StartAsync(CancellationToken cancellationToken = default)
     {
-        if (_publishingOptions.Publisher is not null && _publishingOptions.Publisher != "dcp")
+        if (_publishingOptions.Publisher is not null and not "dcp")
```
Comment (adamint) [inline src/Aspire.Hosting/Dcp/DcpHostService.cs:54]: is there a reason we have a hardcoded string here?

**Review by adamint — APPROVED** [review]

File: src/Aspire.Hosting/Dcp/DcpHostService.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Aspire.Hosting/Dcp/DcpHostService.cs:54]: I've been planning on introducing `KnownPublishers` for the string. It appears in many places throughout the code.

De-duplicating the condition makes sense.

**Comment by drewnoakes** [comment]
> Can we unit test this to stop it regressing again?

The dedupe will keep them in sync. Beyond that, adding a test fixture for this class is lower priority than other items on my plate.

File: src/Aspire.Hosting/Dcp/DcpHostService.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Aspire.Hosting/Dcp/DcpHostService.cs]: https://github.com/dotnet/aspire/pull/1972

**Pull request merged**